File: src/main/java/pe/nanamochi/banchus/entities/db/Score.java
package pe.nanamochi.banchus.entities.db;

import jakarta.persistence.*;
import java.time.Instant;
import lombok.Data;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;
import pe.nanamochi.banchus.entities.commons.Mode;
import pe.nanamochi.banchus.entities.commons.SubmissionStatus;

@Entity
@Data
@Table(
    name = "scores",
    indexes = {
      @Index(
          name = "score_user_mode_status_pp_idx",
          columnList = "user_id, mode, submission_status, performance_points DESC"),
      @Index(name = "beatmap_mode_status_idx", columnList = "beatmap_id"),
      @Index(name = "beatmap_status_idx", columnList = "submissionStatus")
    })
@EntityListeners(AuditingEntityListener.class)
public class Score {
  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  @Column(name = "id", nullable = false)
  private int id;

  @ManyToOne
  @JoinColumn(name = "user_id", nullable = false)
  private User user;

  @Column(name = "online_checksum", length = 32, nullable = false)
  private String onlineChecksum;

  @ManyToOne
  @JoinColumn(name = "beatmap_id", nullable = false)
  private Beatmap beatmap;

  @Column(name = "score", nullable = false)
  private long score;

  @Column(name = "performance_points", nullable = false)
  private double performancePoints;

  @Column(name = "accuracy", nullable = false)
  private double accuracy;

  @Column(name = "highest_combo", nullable = false)
  private int highestCombo;

  @Column(name = "full_combo", nullable = false)
  private boolean fullCombo;

  @Column(name = "mods", nullable = false)
  private int mods;

  @Column(name = "num_300s", nullable = false)
  private int num300s;

  @Column(name = "num_100s", nullable = false)
  private int num100s;

  @Column(name = "num_50s", nullable = false)
  private int num50s;

  @Column(name = "num_misses", nullable = false)
  private int numMisses;

  @Column(name = "num_gekis", nullable = false)
  private int numGekis;

  @Column(name = "num_katus", nullable = false)
  private int numKatus;

  @Column(name = "grade", length = 2, nullable = false)
  private String grade;

  @Column(name = "submission_status", nullable = false)
  private SubmissionStatus submissionStatus;

  @Column(name = "mode", nullable = false)
  private Mode mode;

  @Column(name = "passed", nullable = false)
  private boolean passed;

  @Column(name = "time_elapsed", nullable = false)
  private int timeElapsed;

  @CreatedDate
  @Column(name = "created_at", nullable = false, updatable = false)
  private Instant createdAt;

  @LastModifiedDate
  @Column(name = "updated_at", nullable = false)
  private Instant updatedAt;

  @PrePersist
  @PreUpdate
  public void updateAccuracy() {
    this.accuracy = calculateAccuracy();
  }

  public double calculateAccuracy() {
    if (mode == null) return 0.0;
    return switch (mode) {
      case OSU -> calculateOsuAccuracy();
      case TAIKO -> calculateTaikoAccuracy();
      case CATCH -> calculateCatchAccuracy();
      case MANIA -> calculateManiaAccuracy();
    };
  }

  private double calculateOsuAccuracy() {
    int totalNotes = this.getNum300s() + this.getNum100s() + this.getNum50s() + this.getNumMisses();
    if (totalNotes == 0) return 0.0;

    double accuracy =
        (100.0
            * ((this.getNum300s() * 300.0)
                + (this.getNum100s() * 100.0)
                + (this.getNum50s() * 50.0))
            / (totalNotes * 300.0));
    return clampAccuracy(accuracy);
  }

  private double calculateTaikoAccuracy() {
    int totalNotes = this.getNum300s() + this.getNum100s() + this.getNumMisses();
    if (totalNotes == 0) return 0.0;

    double accuracy = (100.0 * ((this.getNum100s() * 0.5) + this.getNum300s()) / totalNotes);
    return clampAccuracy(accuracy);
  }

  private double calculateCatchAccuracy() {
    int totalNotes =
        this.getNum300s()
            + this.getNum100s()
            + this.getNum50s()
            + this.getNumKatus()
            + this.getNumMisses();
    if (totalNotes == 0) return 0.0;

    double accuracy =
        (100.0 * (this.getNum300s() + this.getNum100s() + this.getNum50s())) / totalNotes;
    return clampAccuracy(accuracy);
  }

  private double calculateManiaAccuracy() {
    int totalNotes =
        this.getNum300s()
            + this.getNum100s()
            + this.getNum50s()
            + this.getNumGekis()
            + this.getNumKatus()
            + this.getNumMisses();
    if (totalNotes == 0) return 0.0;

    double accuracy =
        (100.0
            * ((this.getNum50s() * 50.0)
                + (this.getNum100s() * 100.0)
                + (this.getNumKatus() * 200.0)
                + ((this.getNum300s() + this.getNumGekis()) * 300.0))
            / (totalNotes * 300.0));
    return clampAccuracy(accuracy);
  }

  private double clampAccuracy(double value) {
    return Math.min(100.0, Math.max(0, value));
  }
}

File: banchus-common/src/main/kotlin/pe/nanamochi/banchus/database/entity/Score.kt
package pe.nanamochi.banchus.database.entity

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EntityListeners
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Index
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.PrePersist
import jakarta.persistence.PreUpdate
import jakarta.persistence.Table
import java.time.Instant
import kotlin.math.max
import kotlin.math.min
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import pe.nanamochi.banchus.domain.enums.Mode
import pe.nanamochi.banchus.domain.enums.SubmissionStatus

@Entity
@Table(
    name = "scores",
    indexes =
        [
            Index(
                name = "score_user_mode_status_pp_idx",
                columnList = "user_id, mode, submission_status, performance_points DESC",
            ),
            Index(name = "beatmap_mode_status_idx", columnList = "beatmap_id"),
            Index(name = "beatmap_status_idx", columnList = "submission_status"),
        ],
)
@EntityListeners(AuditingEntityListener::class)
class Score(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id", nullable = false)
    var id: Int = 0,
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    var user: User? = null,
    @Column(name = "online_checksum", length = 32, nullable = false)
    var onlineChecksum: String = "",
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "beatmap_id", nullable = false)
    var beatmap: Beatmap? = null,
    @Column(name = "score", nullable = false) var score: Long = 0L,
    @Column(name = "performance_points", nullable = false) var performancePoints: Double = 0.0,
    @Column(name = "accuracy", nullable = false) var accuracy: Double = 0.0,
    @Column(name = "highest_combo", nullable = false) var highestCombo: Int = 0,
    @Column(name = "full_combo", nullable = false) var fullCombo: Boolean = false,
    @Column(name = "mods", nullable = false) var mods: Int = 0,
    @Column(name = "num_300s", nullable = false) var num300s: Int = 0,
    @Column(name = "num_100s", nullable = false) var num100s: Int = 0,
    @Column(name = "num_50s", nullable = false) var num50s: Int = 0,
    @Column(name = "num_misses", nullable = false) var numMisses: Int = 0,
    @Column(name = "num_gekis", nullable = false) var numGekis: Int = 0,
    @Column(name = "num_katus", nullable = false) var numKatus: Int = 0,
    @Column(name = "grade", length = 2, nullable = false) var grade: String = "F",
    @Column(name = "submission_status", nullable = false)
    @Enumerated(EnumType.STRING)
    var submissionStatus: SubmissionStatus = SubmissionStatus.SUBMITTED,
    @Column(name = "mode", nullable = false)
    @Enumerated(EnumType.ORDINAL)
    var mode: Mode = Mode.OSU,
    @Column(name = "passed", nullable = false) var passed: Boolean = false,
    @Column(name = "time_elapsed", nullable = false) var timeElapsed: Int = 0,
    @CreatedDate
    @Column(name = "created_at", nullable = false, updatable = false)
    var createdAt: Instant = Instant.now(),
    @LastModifiedDate
    @Column(name = "updated_at", nullable = false)
    var updatedAt: Instant = Instant.now(),
) {
    @PrePersist
    @PreUpdate
    fun updateAccuracy() {
        this.accuracy = calculateAccuracy()
    }

    fun calculateAccuracy(): Double =
        when (mode) {
            Mode.OSU -> calculateOsuAccuracy()
            Mode.TAIKO -> calculateTaikoAccuracy()
            Mode.CATCH -> calculateCatchAccuracy()
            Mode.MANIA -> calculateManiaAccuracy()
        }

    private fun calculateOsuAccuracy(): Double {
        val totalNotes = num300s + num100s + num50s + numMisses
        if (totalNotes == 0) return 0.0

        val acc =
            (100.0 * (num300s * 300.0 + num100s * 100.0 + num50s * 50.0) / (totalNotes * 300.0))
        return clampAccuracy(acc)
    }

    private fun calculateTaikoAccuracy(): Double {
        val totalNotes = num300s + num100s + numMisses
        if (totalNotes == 0) return 0.0

        val acc = (100.0 * (num100s * 0.5 + num300s) / totalNotes)
        return clampAccuracy(acc)
    }

    private fun calculateCatchAccuracy(): Double {
        val totalNotes = num300s + num100s + num50s + numKatus + numMisses
        if (totalNotes == 0) return 0.0

        val acc = (100.0 * (num300s + num100s + num50s) / totalNotes)
        return clampAccuracy(acc)
    }

    private fun calculateManiaAccuracy(): Double {
        val totalNotes = num300s + num100s + num50s + numGekis + numKatus + numMisses
        if (totalNotes == 0) return 0.0

        val acc =
            (100.0 *
                (num50s * 50.0 +
                    num100s * 100.0 +
                    numKatus * 200.0 +
                    (num300s + numGekis) * 300.0) / (totalNotes * 300.0))
        return clampAccuracy(acc)
    }

    private fun clampAccuracy(value: Double): Double = min(100.0, max(0.0, value))
}
